[WIP] Add annotations to the individual pages

jwforres opened the pull request.

jwforres commented:
![annotations_collapsed](https://cloud.githubusercontent.com/assets/1467629/10367943/1049c4d2-6da1-11e5-82bb-08f52b69886f.png)
![annotations_expanded](https://cloud.githubusercontent.com/assets/1467629/10367944/104a71d4-6da1-11e5-938f-f2c3251e15a1.png)
![annotations_image_stream_page](https://cloud.githubusercontent.com/assets/1467629/10367942/1047f95e-6da1-11e5-8d1a-738c7643dd2d.png)

jwforres commented:
Used a single directive so it could just be added to every page.  Collapses annotations by default.  The style looks a little out of place on the image stream page though, see screenshots above.

@spadgett

spadgett commented:
Would it be better as just a "Show annotations" link by itself at the bottom of the page? Feels a little odd showing the label but hiding the value, and it might look better for the image streams page. It also wouldn't push other content like pod templates down.

spadgett commented on assets/app/scripts/filters/util.js:
```diff
@@ -296,4 +296,62 @@ angular.module('openshiftConsole')
+  })
+  .filter('isJSON', function() {
+    return function(json) {
+      // return original value if its null or undefined
```
Does this work to avoid duplicate code?

`return !!parseJSONFilter(json);`

jwforres replied:
i'm not even using isJSON anymore, removing it.

spadgett commented on assets/app/scripts/filters/util.js:
```diff
@@ -296,4 +296,62 @@ angular.module('openshiftConsole')
+      // return the parsed obj if its valid
+      try {
+        var jsonObj = JSON.parse(json);
+        if (typeof jsonObj === "object") {
```
Do we need to worry about JSON arrays?

jwforres replied:
yes but they apparently also return typeof as "object"

spadgett replied:
Ah, of course :)

spadgett commented on assets/app/styles/_core.less:
```diff
@@ -645,4 +645,12 @@ select:invalid {
+}
+
+// For when you need formatting of whitespace but don't want all the other pre tag styles
+pre.lightweight {
```
Does it work to simply set `white-space: pre;` on the element? Or possibly `pre-wrap`?

spadgett commented on assets/app/views/browse/pod.html:
```diff
@@ -32,6 +32,7 @@
                   <dd>{{pod.status.podIP || 'unknown'}}</dd>
                   <dt>Restart policy:</dt>
                   <dd>{{pod.spec.restartPolicy || 'Always'}}</dd>
+                  <annotations ng-if="pod.metadata.annotations" annotations="pod.metadata.annotations"></annotations>                  
```
Any reason for the `ng-if` but not elsewhere?

jwforres replied:
Carryover from code changes, it can be removed, the annotations directive handles it.

jwforres commented:
Switched to "Show annotations" at the bottom of every page.  I think the image stream page needs a revamp anyway, added a specific item to the punch list for that.

[merge]
